sycl/include/CL/sycl/queue.hpp

```diff
@@ -355,36 +355,80 @@ class __SYCL_EXPORT queue {
     return submit([=](handler &CGH) { CGH.prefetch(Ptr, Count); });
   }
 
+// having _TWO_ mid-param #ifdefs makes the functions very difficult to read.
+// Here we simplify the &CodeLoc declaration to be CODELOCPARAM(A,B,C,D,E) and
+// CODELOCARG(B,C,D) Similarly, the KernelFunc param is simplified to be
+// KERNELFUNCPARAM(B,F,G) Once the queue kernel functions are defined, these
+// macros are #undef immediately. Compare this first definition of single_task
+// here, with the second one below. replace CODELOCPARAM(A,B,C,D,E) with nothing
+// or :   , const detail::code_location &CodeLoc =
+// detail::code_location::current() replace CODELOCARG(B,C,D) with nothign or :
+// const detail::code_location &CodeLoc = {};
+#define A ,
+#define B const
+#define C detail::code_location
+#define D &CodeLoc
+#define E detail::code_location::current()
+#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
+#define CODELOCPARAM(a, b, c, d, e) a b c d = e
+
+#define CODELOCARG(b, c, d)
+#else
+#define CODELOCPARAM(a, b, c, d, e)
+
+#define CODELOCARG(b, c, d) b c d = {}
+#endif
+// replace KERNELFUNCPARAM(B,F,G) with   KernelType KernelFunc
+//                                  or   const KernelType &KernelFunc
+#define F KernelType
+#define G KernelFunc
+#ifdef __SYCL_NONCONST_FUNCTOR__
+#define KERNELFUNCPARAM(B, F, G) F G
+#else
+#define KERNELFUNCPARAM(B, F, G) B F &G
+#endif
+
   /// single_task version with a kernel represented as a lambda.
   ///
   /// \param KernelFunc is the Kernel functor or lambda
   /// \param CodeLoc contains the code location of user code
   template <typename KernelName = detail::auto_name, typename KernelType>
-  event single_task(
-      KernelType KernelFunc
-#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA
-      ,
-      const detail::code_location &CodeLoc = detail::code_location::current()
-#endif
-  ) {
-#ifdef DISABLE_SYCL_INSTRUMENTATION_METADATA
-    const detail::code_location &CodeLoc = {};
-#endif
+  event single_task(KERNELFUNCPARAM(B, F, G) CODELOCPARAM(A, B, C, D, E)) {
+    CODELOCARG(B, C, D);
+
```

Comment:
This declaration of single_task is much easier to read than the next one, by virtue of these macros.  @bader  , @romanovvlad let me know your thoughts. I can remove them, or use them in the other affected declarations in queue.hpp.